llvm/lib/SYCL/InSPIRation/InSPIRation.cpp

```diff
@@ -337,6 +338,79 @@ struct InSPIRation : public ModulePass {
     }
   }
 
+  /// This currently looks through the arguments passed to the SSDM intrinsic
+  /// call and checks the instruction to see if it is an address space cast to
+  /// a generic, if it is, it will take the concrete segment of the cast and
+  /// replace the operand with it. It looks like the below example:
+  ///
+  /// Before:
+  /// %10 = getelementptr inbounds %"struct._ZTSN2cl4sycl6xilinx15partition_
+  ///   arrayIcLm9ENS1_9partition8completeILm0EEEEE.cl::sycl::xilinx::partition_
+  ///   array", %"struct._ZTSN2cl4sycl6xilinx15partition_arrayIcLm9ENS1_9
+  ///   partition8completeILm0EEEEE.cl::sycl::xilinx::partition_array"* %1,
+  ///   i64 0, i32 0, i64 0
+  /// %11 = addrspacecast i8* %10 to i8 addrspace(4)*
+  /// call spir_func void (...) @_ssdm_SpecArrayPartition(i8 addrspace(4)* %11,
+  ///   i64 0, i8* getelementptr inbounds ([9 x i8], [9 x i8]* @.str, i64 0,
+  ///   i64 0), i32 0, i8* getelementptr inbounds ([1 x i8], [1 x i8]* @.str.2,
+  ///   i64 0, i64 0)) #4
+  ///
+  /// After:
+  /// %10 = getelementptr inbounds %"struct._ZTSN2cl4sycl6xilinx15partition_
+  ///   arrayIcLm9ENS1_9partition8completeILm0EEEEE.cl::sycl::xilinx::partition
+  ///   _array", %"struct._ZTSN2cl4sycl6xilinx15partition_arrayIcLm9ENS1_9
+  ///   partition8completeILm0EEEEE.cl::sycl::xilinx::partition_array"* %1,
+  ///   i64 0, i32 0, i64 0
+  ///  call spir_func void (...) @_ssdm_SpecArrayPartition(i8* %10, i64 0,
+  ///   i8* getelementptr inbounds ([9 x i8], [9 x i8]* @.str, i64 0, i64 0),
+  ///   i32 0, i8* getelementptr inbounds ([1 x i8], [1 x i8]* @.str.2, i64 0,
+  ///   i64 0)) #4
+  ///
+  /// It's simply collapsing away the cast right now, this doesn't consider the
+  /// possibility of interactions with other possible address space casts that
+  /// depend on it (we simply replace all uses with the non-geneirc variant).
+  /// For now we hope that these are erased or eliminated by either of the AS
+  /// Fixer passes or DCE passes.
+  ///
+  /// Note: Unsure how robust this is with the small sample size we currently
+  /// have to test against. So it's a WIP. If the situation becomes untenable
+  /// we can likely revert the accessor class back to it's prior form by
+  /// reverting this pull: https://github.com/intel/llvm/pull/348 (commit:
+  /// 609999c4e1aeca05aff010ce5e2eb08dde08fd69). This may cause address space
+  /// leakage however, but should result in more overall address space
+  /// consistency/stability due to the addition of more concrete address spaces.
+  void handleSpecArrayPartition(CallInst *CI) {
+    for (unsigned int i = 0; i < CI->getNumOperands(); ++i) {
+      if (CI->getOperand(i)->getType()->isPointerTy()) {
+        if (auto *ASC = dyn_cast<AddrSpaceCastInst>(CI->getOperand(i))) {
+          if (ASC->getDestAddressSpace() == /*Generic Address Space*/4) {
```

Comment:
That argument goes to a different pass, but I could add it as part of InSPIRation as well and see if it will pass it into here, so it will stay synchronized.

Reply:
Seems like you cannot register the same named argument for multiple passes and the method to share pass options across pass boundaries seems a little intrusive (the InferAddressSpace pass also doesn't have a header which makes the change a little weirder): http://llvm.org/docs/CommandLine.html#internal-vs-external-storage

This seems like it would be the ideal thing to use, but it doesn't seem to have any guarantees that the options will be registered when we need them and the type it returns doesn't really seem to allow access to the value in any case:
http://llvm.org/docs/CommandLine.html#the-cl-getregisteredoptions-function

I could register another argument with a different name for this pass that will accept an identical value or we could leave it as-is for now?